email_validator/__init__.py

```diff
@@ -470,6 +478,8 @@ def dns_resolver_resolve_shim(resolver, domain, record):
             raise dns.exception.Timeout()
 
         resolver = dns.resolver.get_default_resolver()
+        if resolver.cache is None and dns_resolver_cache is not None:
+            resolver.cache = dns_resolver_cache
```

Comment:
This modifies the global default resolver, which isn't great. It looks like we have to instantiate our own resolver instance and set its cache field to avoid that. I'm not sure if instantiating a new resolver has a performance penalty --- possibly, since it will read /etc/resolv.conf for the nameserver to use (on linux at least). If that's the case, the right way to do this might be to have the caller instantiate and pass a resolver (and be responsible for setting the cache field) rather than a cache class. That might be more powerful anyway.

Reply:
I now see we already modify the global default resolver right below your change for the timeout parameter. We really shouldn't be modifying global state. So we could fix both problems with one change. The only problem would be that we'd lose support for the timeout parameter, which would be a breaking change. But I'm OK with that.

Reply:
Sorry for lots of comments. Alternatively, we could try cloning the default resolver before modifying it.

Reply:
Thanks for the comments.

Yeah I'm not sure if instantiating a new resolver + reading `/etc/resolve.conf` adds noticeable overhead or not.

Hm so maybe we can keep a module local resolver instance instead. Plus have a optional use`configure_dns_resolver()` function where one could modify timeout + dns cache. I can update this PR to show how I mean.

Also what do you think about using cache by default?

Reply:
I'm generally opposed to any sort of global variable or global state, which makes a default cache impossible.

I believe the right approach here is the second one I mentioned - let the caller instantiate the resolver. We can add some documentation for how the user should do this. Or we can add a helper method to construct a resolver with a timeout and cache.

Reply:
I agree avoiding global variables and state is generally good. But it also feels kind of strange API-wise to force users of email validator keeping track of a dns resolver instance 🤔

Maybe keeping `get_default_resolver()` from dns package can work in the default case, but if a resolver is provided we use that.

I just force pushed this change. Let me know what you think @JoshData 
Note that the code is still backwards compatible, since I've kept the timeout arg.